daemon/cmd/fqdn.go

```diff
@@ -131,6 +131,10 @@ func (d *Daemon) bootstrapFQDN(possibleEndpoints map[uint16]*endpoint.Endpoint,
 		Cache:           fqdn.NewDNSCache(option.Config.ToFQDNsMinTTL),
 		UpdateSelectors: d.updateSelectors,
 	}
+	// Disable cleanup tracking on the default DNS cache. This cache simply
+	// tracks which api.FQDNSelector are present in policy which apply to
+	// locally running endpoints.
+	cfg.Cache.DisableCleanupTrack()
```

Comment:
I read some of this code for the first time in the last week so I'm not fully familiar with it yet. But the fix seems a bit weird to me - isn't there a better way to handle fix this rather than disabling writes to the cache, thereby GC as the PR changes suggest?

Reply:
> isn't there a better way to handle fix this rather than disabling writes to the cache, thereby GC as the PR changes suggest?

I don't understand the question nor what is being proposed. Bare in mind this fix needs to be minimal to avoid regressions in backports to older branches. Feel free to propose a different fix that does not need to be backported as a follow up PR.

Reply:
Why don't we run `GC` on the global cache if we are still continuing to keep the cache around similar to how it's run on the per-endpoint cache? At the very least, is there a reason why we don't block the writes to the cache at the entry point (e.g., [ReplaceFromCacheByNames](https://github.com/cilium/cilium/blob/master/pkg/fqdn/cache.go#L374))? The current fix is rendering most of this code as dead code.

Reply:
> Why don't we run `GC` on the global cache if we are still continuing to keep the cache around similar to how it's run on the per-endpoint cache? 

It's not necessary because the Cache is replaced here https://github.com/cilium/cilium/blob/fcc345ec0a00ad4a385f650910f1d35e653a1a2e/daemon/cmd/fqdn.go#L220

> At the very least, is there a reason why we don't block the writes to the cache at the entry point (e.g., [ReplaceFromCacheByNames](https://github.com/cilium/cilium/blob/master/pkg/fqdn/cache.go#L374))? The current fix is rendering most of this code as dead code based on my limited understanding of the code.

I'm not following, block which writes?

Reply:
Well, I was referring to this check added by the PR - 

```
func (c *DNSCache) addNameToCleanup(entry *cacheEntry) {
	if c.cleanup == nil {
		return
	}
```

Some of the code is shared between global and per-endpoint cache updates so teasing out the clean-up tracking related code can complicate backports. We can go with this minimal fix as you mentioned about keeping the changes minimal for backports.